### PR TITLE
Dev opt inference

### PR DESCRIPTION
**Pull request by CPFLAME**
这个PR要做的:

- [x] 使用mock功能来支持导入transformers仓库
- [x] 单机单卡的opt推理
- [x] 单机多卡的opt 模型并行推理
- [x] 和hugging face的正确性验证

文档说明:

# Mock transformers

This is an application of mock [transformers](https://github.com/huggingface/transformers), which can perform distributed inference in LiBai with model under the transformers.

**Supported Model**

- [OPT](#distributed-infer-opt): tensor parallel

## Environment 

Before running the scripts, make sure to install the library's dependencies:

### Install libai

libai installation, refer to [Installation instructions](https://libai.readthedocs.io/en/latest/tutorials/get_started/Installation.html)

```bash
# create conda env
conda create -n libai python=3.8 -y
conda activate libai
# install oneflow nightly, [PLATFORM] could be cu117 or cu102
python3 -m pip install --pre oneflow -f https://staging.oneflow.info/branch/master/[PLATFORM]
# install libai
git clone https://github.com/Oneflow-Inc/libai.git
cd libai
pip install pybind11
pip install -e .
```

- All available `[PLATFORM]`:

    <table class="docutils">
    <thead>
    <tr class="header">
    <th>Platform</th>
    <th>CUDA Driver Version</th>
    <th>Supported GPUs</th>
    </tr>
    </thead>
    <tbody>
    <tr class="odd">
    <td>cu117</td>
    <td>&gt;= 450.80.02</td>
    <td>GTX 10xx, RTX 20xx, A100, RTX 30xx</td>
    </tr>
    <tr class="even">
    <td>cu102</td>
    <td>&gt;= 440.33</td>
    <td>GTX 10xx, RTX 20xx</td>
    </tr>
    <tr class="odd">
    <td>cpu</td>
    <td>N/A</td>
    <td>N/A</td>
    </tr>
    </tbody>
    </table></li>

### Install transformers

refer to [transformers installation](https://github.com/huggingface/transformers#installation)
```
python3 -m pip install "transformers>=4.26"
```

Notes

- You need to register a Hugging Face account token and login with `huggingface-cli login`

```bash
python3 -m pip install huggingface_hub
```

- If no command available in the PATH, it might be in the `$HOME/.local/bin`

```bash
 ~/.local/bin/huggingface-cli login
```

## distributed infer OPT

An reimplement of [OPT](https://github.com/huggingface/diffusers/tree/main/examples/dreambooth) distributed inference in LiBai

<table class="docutils">
  <tbody>
    <tr>
      <th width="80"> opt inference </th>
      <th valign="bottom" align="left" width="120">Tensor Parallel</th>
      <th valign="bottom" align="left" width="120">Pipeline Parallel</th>
    </tr>
    <tr>
      <td align="left"> <b> Support </b> </td>
      <td align="left">&#10004;</td>
      <td align="left">-</td>
    </tr>
  </tbody>
</table>

for `tensor_parallel=2`, run command in `libai_root`
```
bash tools/infer.sh projects/mock_transformers/dist_infer_opt.py 2
```
modify the infer code `dist_infer_opt.py` according to your own needs:
```python
...
if __name__ == "__main__":
    # set dist config
    parallel_config = DictConfig(
        dict(
            data_parallel_size=1,
            tensor_parallel_size=2, # modify it according to your own needs
            pipeline_parallel_size=1, # set to 1, unsupport pipeline parallel now
            pipeline_num_layers=None,
            )
    )
    dist.setup_dist_util(parallel_config)
    ...
    # initial and load model
    model = AutoModelForCausalLM.from_pretrained("facebook/opt-125m").half() # change your model type 125m~66b
    model._apply(dist.convert_to_distributed_default_setting)
    # initial tokenizer
    tokenizer = AutoTokenizer.from_pretrained("facebook/opt-125m", use_fast=False) # change your model type  125m~66b
```

### CODE REVIEW

**Comment by xiezipeng-ML** [comment]
阔以加一下License，然后init_env.py的潜力很大，感觉之后可以mock很多任务来兼容

File: projects/mock_transformers/init_env.py [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+import oneflow as flow
+
+flow.mock_torch.enable()
+
+from oneflow import Tensor, nn  # noqa
+from transformers import modeling_utils  # noqa
+from transformers.modeling_utils import _load_state_dict_into_model  # noqa
+
+# ---------------- mock _load_state_dict_into_model ------------------
+new_load = _load_state_dict_into_model
```
Comment (Ldpe2G) [inline projects/mock_transformers/init_env.py:11]: 这一句需要加么，因为我看下面 new_load 的实现没有再引用原来的 `_load_state_dict_into_model` 实现了

File: projects/mock_transformers/dist_infer_opt.py [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+        sbp=dist.get_nd_sbp([flow.sbp.broadcast, flow.sbp.broadcast]),
+        placement=dist.get_layer_placement(0),
+    )
+
+    # generate id
+    placement_sbp_dict = dict(
+        placement=flow.env.all_device_placement("cuda"),
+        sbp=flow.sbp.broadcast,
+    )
+    with global_mode(True, **placement_sbp_dict):
```
Comment (Ldpe2G) [inline projects/mock_transformers/dist_infer_opt.py:74]: 这里加 global mode 的意义是？前面整个模型都已经是 to_global 了，输入也是 global tensor

File: projects/mock_transformers/init_env.py (hunk 1 above)
Comment (CPFLAME) [reply projects/mock_transformers/init_env.py:26]: 不用了 我把它删了

File: projects/mock_transformers/dist_infer_opt.py (hunk 2 above)
Comment (CPFLAME) [reply projects/mock_transformers/dist_infer_opt.py:74]: 这里是因为 在model.generate(也可以说是model.forward())里面有大量的local语句, 比如什么 .device(), torch.ones()之类的, 为了在推理的时候也使用global tensor, 所以要在这里添加

**Comment by CPFLAME** [comment]
> 阔以加一下License，然后init_env.py的潜力很大，感觉之后可以mock很多任务来兼容

加了

File: projects/mock_transformers/dist_infer_opt.py (hunk 2 above)
Comment (Ldpe2G) [reply projects/mock_transformers/dist_infer_opt.py:74]: 懂了

**Review by xiezipeng-ML — APPROVED** [review]

**Review by Ldpe2G — APPROVED** [review]

**Pull request merged**